packages/x-bus/src/x-bus.ts

```diff
@@ -0,0 +1,245 @@
+ *
+ * @public
+ */
+export class XPriorityBus implements XBus {
+  /**
+   *
+   *
+   * @internal
+   */
+  protected queue: XPriorityQueue<XEvent>;
+
+  /**
+   *
+   *
+   * @internal
+   */
+  protected priorities: Dictionary<number>;
+
+  /**
+   *
+   *
+   * @internal
+   */
+  protected emitCallbacks: AnyFunction[];
+
+  /**
+   * A dictionary to store the created event emitters.
+   *
+   * @internal
+   */
+  protected emitters: Emitters = {};
+
+  /**
+   * A pending flush operation {@link TimeoutId | timeout identifier} or undefined if there's
+   * none pending.
+   *
+   * @internal
+   */
+  protected pendingFlush?: TimeoutId;
+
+  /**
+   * A list of pending pop operations {@link TimeoutId | timeout identifiers}.
+   *
+   * @internal
+   */
+  protected pendingPops: TimeoutId[] = [];
+
+  /**
+   * Creates a new instance of a {@link XPriorityBus}.
+   *
+   * @param config
+   * @param queue - A {@link XPriorityQueue} to store the events.
+   * @param priorities - A {@link @empathyco/x-utils#Dictionary} defining the priorities associated
+   * to a given string.
+   */
+  public constructor(
+    config: {
+      queue?: XPriorityQueue;
+      priorities?: Dictionary<number>;
+      emitCallbacks?: AnyFunction[];
+    } = {}
+  ) {
+    this.queue = config.queue ?? new XPriorityQueue();
+    this.priorities = config.priorities ?? {
+      RequestChanged$: 2,
+      Changed$: 4,
+      '^User': 6,
+      FromUrl$: 8,
+      '^External': 10,
+      Provided$: 12,
+      Hook$: 12,
+      Received$: 14
+    };
+    this.emitCallbacks = config.emitCallbacks ?? [];
+  }
+
+  /**
+   * Emits an event. See {@link XBus.(emit:2)}.
+   *
+   * @param event - Event to be emitted.
+   * @param payload - Event payload.
+   * @param metadata - Extra event data.
+   *
+   * @returns A promise that is resolved whenever the emitted is truly emitted.
+   */
+  emit<Event extends XEvent>(
+    event: Event,
+    payload?: XEventPayload<Event>,
+    metadata: Dictionary = { moduleName: null }
+  ): Promise<EmittedData<XEvent>> {
+    return new Promise(resolve => {
+      this.queue.push(event, this.getEventPriority(event, metadata), {
+        replaceable: metadata.replaceable || false,
+        resolve,
+        eventPayload: payload,
+        eventMetadata: metadata
+      });
+
+      this.flushQueue();
+    });
+  }
+
+  /**.
+   * Retrieves the event priority. The criteria to get the priority is:
+   * - the defined event metadata priority
+   * - the priority associated to the matching preconfigured priority key
+   * - the max positive value is assigned
+   *
+   * @param event - The {@link XEvent | event} to get the priority from.
+   * @param metadata - The {@link XEvent | event} metadata.
+   * @returns The priority for the given {@link XEvent | event}.
+   *
+   * @internal
+   */
+  getEventPriority(event: XEvent, metadata: Dictionary): Priority {
+    const [, priorityKey] = Object.keys(this.priorities).reduce(
```

Comment:
Because of how this code is made, we are deciding which priority is more refined using the length of the matched part. For example, if you emit an event 'UserRequestChanged' if you don't compare which priority is more refined, the priority will be 
 `Changed` instead of `RequestChanged`
 
After a discussion, we decided to not do this. It doesn't make sense to check if a 'matched' key is more refined than another one, to avoid that, we should change the name of the `RequestChanged`events to   `RequestUpdated`  or create a more restricted regex. In this case, we should rename the events to be `RequestUpdated`

Apart from this:

Reply:
Refactored based on discussion